store/store.go

```diff
@@ -261,24 +264,66 @@ func (s *Store) GetModel(name string) *Model {
 
 // Reduce processes txn events into the models.
 func (s *Store) Reduce(events []core.Event) error {
-	codecActions, err := s.eventcodec.Reduce(events, s.datastore, baseKey)
+	txn, err := s.datastore.NewTransaction(false)
 	if err != nil {
 		return err
 	}
-	actions := make([]Action, len(codecActions))
-	for i, ca := range codecActions {
+	defer txn.Discard()
+
+	actions := make([]Action, len(events))
```

Comment:
So now all the datastore-based logic is in here, rather than in the EventCodec. This is an initial step towards making it easier to do things like indexing, without requiring that a developer implements indexing within the EventCodec (which didn't feel right).

Reply:
I _think_ the intention was to put indexing outside of the codec, right @carsonfarmer? I don't think these changes here are needed for that work, unless I am misunderstanding...

Reply:
Well, the reason it seemed like the indexing would need to happen in the Codec was that to index the data, we need access to the 'raw' data. In theory, only the Model and Codec really know anything about that data. We only want to index valid data, so it has to happen _in_ or _after_ the Codec has done its work. It didn't seem like a good place to have _in_ the codec (as you point out), but ideally the data should be indexed in the same transaction that is being used to add the 'raw' data to the datastore. This currently happens in the Codec.
As a side note, it wasn't clear to me _why_ the Codec should need to know anything about the datastore per se in the first place? Hence pulling out that transaction logic into the Store in the proposal.

Reply:
The raw data could be retrieved directly from the datastore. I can't remember if it's CBOR or JSON. Maybe it depends on JSON mode. In any case, I don't see why it couldn't be loaded and indexed in a generic way after the codec has done it's job.

Reply:
Agreed, it does seem like this is the right approach. I'm not proposing the indexing happens in the Codec, I'm proposing it be done outside as you suggest. What I _am_ proposing here is that, it could be done in the same transaction (which it should) if we move that transaction logic _out_ of the Codec and _into_ the Store. Make sense?

Reply:
(Also yes you are right, it depends on the 'mode'.)

Reply:
Ok, cool. Sounds good 👍

Reply:
> As a side note, it wasn't clear to me _why_ the Codec should need to know anything about the datastore per se in the first place? Hence pulling out that transaction logic into the Store in the proposal.

I think most of the discussion relates on how much responsibility and power is given to `EventCodec`s. From my POV, the event codec should transform entity changes in a txn, to whatever the EventCodec decides it's the best way to express it and apply that to the model state (hence, works on the datastore too).

With your change, the EventCodec returns the new state to the Store and ask the Store to execute that change in the datastore. So, it will _reply_ something like: `Create entity with this bytes`. What I think of this is that it adds more constraints to what the EventCodec can do since it must frame every result on returning: bytes + (Create || Delete || Update).

In the same vein, part of the idea of the Reducer receive a slice of Events instead of one-by-one (as in your change) is that it may allow the EventCodec to apply multiple changes of the same txn more efficiently (maybe it won't need a datastore txn all the way, now it's forced). Also, an EventCodec could decide to create a single operation that modifies multiple entities (since it compressed txn changes in a single codec operation)... and this can't be possible now since it must return: bytes + (Create || Delete || Update) for a *single* entity.

As I said earlier, I'm not against reducing the scope/power of what EventCodecs can do. But just trying to mention tradeoffs.

Reply:
All fair points. Based on our offline discussion, we'll be taking a different approach. Though one thing is, actually the only thing the EventCodec really _needs_ to return is the new state (and the action if we want). It doesn't need to specify _what_ to do with the state, the store just takes it, and applies it. If the state is nil, it deletes it, otherwise, it overwrites it. The Store then has no knowledge of _what_ the EventCodec did... so no new constraints are required (note that's not how I implemented it to keep things simple, but it could be that way easily). This _does_ however necessitate that each event is handled in turn, but that's currently how it works anyway so...

Reply:
Well, that actually would take out the double-check that was done about not processing two times a `Create` event (which should be fatal). Now it won't complain (since can't distinguish between updates and create, if I'm understanding correctly).
That kind of thing shouldn't happen, since is a sign that something quite bad happened (eventcodec work incorrectly; applying events to corrupted state; threadservice duplicating records).

Actually, this was the way that I realized the threadservice had all the concurrency things I found (since it was panicking for that case).

Reply:
Not sure I follow? If a create event comes thru, the event codec would check if its valid (has existing state), if valid, new state is returned and 'put' by store, if not, throw an error. For a save event, if valid (has existing state), return updated state, if not, same deal as above. Then for delete, return nil as the new state (probs don't need to throw if it doesn't exist). Maybe I'm missing something else?

Reply:
So both are accessing the store?
I think there might be some confusion regarding some applies and rollback changes through discussions; so maybe I'm the one confused. I mainly focused on this claim:

> it doesn't need to specify what to do with the state, the store just takes it, and applies it. If the state is nil, it deletes it, otherwise, it overwrites it.

If the EventCodec is doing the check, then shouldn't be a problem.

Reply:
Yes, the EventCodec is doing the check. But I think here's the source of confusion for us both: it doesn't need access to the store if each Event is processed in turn. But if we did want to keep them being processed in batch (so the EventCodec could so some clever stuff?), then it seems like it would need access to the batch at least...
So in the end, it is as you say: we're talking about two approaches to the same thing, and until we have something specific that needs changing, its just a matter of opinions I think 😄.

Reply:
I would prefer to just leave stuff as it is unless it's needed by some indexer, given that is the goal of this work, not refactoring.